### PR TITLE
Introduce decorator syntax for event listeners

### PR DESCRIPTION
**Pull request by aliabid94**
Added the ability to attach event listeners via decorators, e.g.

```python
with gr.Blocks() as demo:
    name = gr.Textbox(label="Name")
    output = gr.Textbox(label="Output Box")
    greet_btn = gr.Button("Greet")

    @greet_btn.click(inputs=name, outputs=output)
    def greet(name):
        return "Hello " + name + "!"
```

I think this syntax will be much more intuitive and maintainable, especially because the arguments of the function will more clearly correspond to their respective components since they are one line away from each other.

### CODE REVIEW

**Comment by abidlabs** [comment]
Nice @aliabid94! Should we merge into `main` or keep on a separate `dev-4.0` branch? I'm inclined towards merging into `main` but not "popularizing" this too much until we release 4.0

**Comment by aliabid94** [comment]
I actually think we can just merge to main and also publicize, it's not a big change just some syntactic sugar

File: demo/blocks_flipper/run.py
Comment (abidlabs) [inline demo/blocks_flipper/run.py:1]: We shouldn't edit the `blocks_flipper` demo as it is used in the quick start (before the decorator syntax is introduced)

File: .changeset/fuzzy-numbers-repair.md [hunk 1]
```diff
@@ -0,0 +1,5 @@
+---
+"gradio": minor
+---
+
+feat:Introduce decorator syntax for event listeners
```
Comment (abidlabs) [inline .changeset/fuzzy-numbers-repair.md:5]: Can we make this a highlight?

File: gradio/blocks.py [hunk 2]
```diff
@@ -1626,7 +1626,7 @@ def get_time():
                 every=every,
                 no_target=True,
             )
-            return Dependency(self, dep, dep_index)
```
Comment (abidlabs) [inline gradio/blocks.py:1629]: Need to handle the the decorator case for `Blocks.load()` as well, right?

File: gradio/events.py [hunk 3]
```diff
@@ -90,7 +95,7 @@ def __init__(
 
     def __call__(
         self,
-        fn: Callable | None,
+        fn: Callable | None | Literal["wrapper"] = "wrapper",
```
Comment (abidlabs) [inline gradio/events.py:98]: Just wondering, where is this "wrapper" coming from? I don't see it anywhere else in the code.

**Review by abidlabs — APPROVED** [review]

File: gradio/blocks.py (hunk 2 above)
Comment (aliabid94) [reply gradio/blocks.py:1629]: the Blocks.load method is such a nasty mess right now, we can add this later once we fix the load method in 4.0.

File: gradio/events.py (hunk 3 above)
Comment (aliabid94) [reply gradio/events.py:98]: it's just the default value - if you don't set it, it acts like a decorator. There's no way for a function to know if it's being used as a decorator, so if you don't set fn, we assume it's a decorator. Previously fn had no default value so this is not a breaking change.

File: demo/blocks_flipper/run.py
Comment (aliabid94) [reply demo/blocks_flipper/run.py:1]: ok

File: gradio/events.py (hunk 3 above)
Comment (aliabid94) [reply gradio/events.py]: also renamed to "decorator" to be clearer

**Comment by aliabid94** [comment]
@abidlabs addressed everything

**Comment by abidlabs** [comment]
The syntax for the `highlight` changelog isn't quite correct. You have to use `####` for the heading

**Comment by abidlabs** [comment]
Like this: https://github.com/gradio-app/gradio/pull/5279/files

Otherwise looks good!

**Comment by abidlabs** [comment]
cool should be good to merge now

**Comment by FarukOzderim** [comment]
Cool and convenient API and DX @aliabid94, like it!

**Pull request merged**